### ISSUE TITLE
Configurable Maximum `chainHead` Subscriptions for WebSocket Connections in JSON-RPC Server

### ISSUE THREAD

**Issue by josepot** [paritytech/polkadot-sdk#6758]
It's a common practice among RPC providers to place the PolkadotSDK node (serving the JSON-RPC API) behind a reverse proxy. This setup requires WebSocket connections to be multiplexed effectively. However, the current implementation of the JSON-RPC server makes this infeasible because it strictly limits each WS connection to **two `chainHead` subscriptions** [reference code](https://github.com/paritytech/polkadot-sdk/blob/896c81440c1dd169bd2f5e65aba46eca228609f8/substrate/client/rpc-spec-v2/src/chain_head/subscription/inner.rs#L1350): https://github.com/paritytech/polkadot-sdk/blob/896c81440c1dd169bd2f5e65aba46eca228609f8/substrate/client/rpc-spec-v2/src/chain_head/subscription/inner.rs#L1350

### Problem
This hard limit creates significant challenges in scaling and reliability:
1. **Reverse Proxy Compatibility**: The inability to configure the maximum number of `chainHead` subscriptions per WS connection prevents effective multiplexing of "internal" connections behind the reverse proxy. This limitation forces the use of custom proxy-enhancers to deploy the PolkadotSDK node behind a reverse proxy.
   
2. **High Traffic Impact**: When traffic increases, DApps experience severe difficulties establishing `chainHead` subscriptions, often resulting in failures to connect for extended periods (sometimes several minutes). This significantly degrades user experience and operational reliability.

### Current Mitigation and its Limitations
At PAPI, we've developed a custom solution to partially mitigate this issue:
- When a `chainHead` subscription is rejected (without having reached the theoretical limit that the spec guarantees), our client switches to the next available RPC URI. 
- Even if there’s only one configured URI, reconnection can succeed because the reverse proxy may assign the client to a different internal connection with available `chainHead` subscription slots.

While this workaround helps in some cases, it doesn't fully solve the problem:
- In high-load scenarios, reconnection attempts frequently fail, making it impossible to establish a subscription for extended periods.
- This leads to disruptions in DApps, especially during traffic spikes.

### Impact and Urgency
This issue is highly pressing, as it directly impacts the stability and scalability of DApps using the new JSON-RPC APIs. Without a resolution, RPC providers must rely on brittle, custom mitigations that fail to guarantee consistent behavior under high traffic.

**Comment by jsdw** [comment]
Yup agreed; the spec only mandates that a minimum of two chainHead_follow subscriptions be allowed on a connection, but no reason to not allow many more if configured.

> At PAPI, we've developed a custom solution to partially mitigate this issue

I guess you'll still need to keep this code, but the frequency that you need to do something will hopefully decrease.

A smarter reverse proxy would also help (though again, unless the subscription limit is exposed somewhere, it must assume a maximum of 2 follow subscriptions will succeed and so would also have to either stick to this limit regardless, or do something like what PAPI does re trying and establishing new connection on failure, or one connection per client instead of multiplexing)

**Comment by mfornos** [comment]
> A smarter reverse proxy would also help (though again, unless the subscription limit is exposed somewhere, it must assume a maximum of 2 follow subscriptions will succeed and so would also have to either stick to this limit regardless, or do something like what PAPI does re trying and establishing new connection on failure, or one connection per client instead of multiplexing)

Even if the maximum number of subscriptions is exposed, the reverse proxy would still need to manage and enforce this limit.

For reference, just in case someone is interested, HAProxy does not multiplex websocket connections by default, they are marked as "private".

> Persistent connections between HAProxy and the server are flagged as “private” in the following cases:
> * NTLM authentication is active between the client and the server. NTLM authenticates the TCP connection. So of course, this connection between the Proxy and the server can’t be shared between multiple clients.
> * Websocket: only the client who triggered the “upgrade” can speak over this connection.
> * [Transparent proxying](https://www.haproxy.com/glossary/what-is-a-transparent-proxy): in this case, the proxy spoofs the client IP address to get connected to the server.
>
> Even though HAProxy supports persistent connections on the server side, it will follow transparently and safely the principle of “private” connections for the cases described above.
>
>[...]
>
> Connection reuse is managed by placing an [http-reuse directive](https://cbonte.github.io/haproxy-dconv/2.3/configuration.html#http-reuse) in a backend section, which can take the following values:
> * never: all server side connections are considered “private” and HAProxy will never share a connection between multiple clients. 
> * safe (default): the first request of a client is always sent over a dedicated connection for the client. Subsequent requests from the client can be sent over existing idle connections.
> * aggressive: the first request of a client can be sent over any connection that has been used at least once, providing the server properly supports persistent connections. Of course, a new connection may be opened in case no idle ones match the requirement described previously.
> * always: the first request of a client is always sent on an existing idle connection. A new connection may be established if none of them are available.

From: https://www.haproxy.com/blog/http-keep-alive-pipelining-multiplexing-and-connection-pooling
